File: argon-compiler-core/shared/src/main/scala/dev/argon/compiler/output/ArgonModuleSerialized.scala
package dev.argon.compiler.output

import dev.argon.compiler._
import dev.argon.{module => ArgonModule}

trait ArgonModuleSerialized {

  val metadata: ArgonModule.Metadata
  def references: Comp[ArgonModule.ModuleReferencesList]

  def namespaces: CompStream[ArgonModule.NamespaceDeclaration]
  def namespaceElements(id: Int): CompStream[ArgonModule.GlobalDeclarationElement]

  def getTraitDef(id: Int): Comp[ArgonModule.TraitDefinition]
  def getTraitRef(id: Int): Comp[ArgonModule.TraitReference]

  def getClassDef(id: Int): Comp[ArgonModule.ClassDefinition]
  def getClassRef(id: Int): Comp[ArgonModule.ClassReference]

  def getDataConstructorDef(id: Int): Comp[ArgonModule.DataConstructorDefinition]
  def getDataConstructorRef(id: Int): Comp[ArgonModule.DataConstructorReference]

  def getFunctionDef(id: Int): Comp[ArgonModule.FunctionDefinition]
  def getFunctionRef(id: Int): Comp[ArgonModule.FunctionReference]
  def getFunctionPayload(id: Int, platformId: String): CompStream[Byte]

  def getMethodDef(id: Int): Comp[ArgonModule.MethodDefinition]
  def getMethodRef(id: Int): Comp[ArgonModule.MethodReference]
  def getMethodPayload(id: Int, platformId: String): CompStream[Byte]

  def getClassConstructorDef(id: Int): Comp[ArgonModule.ClassConstructorDefinition]
  def getClassConstructorRef(id: Int): Comp[ArgonModule.ClassConstructorReference]
  def getClassConstructorPayload(id: Int, platformId: String): CompStream[Byte]

}

File: argon-plugin-api/src/main/java/dev/argon/plugin/api_v1/ArgonModuleSerialized.java
package dev.argon.plugin.api_v1;

import dev.argon.module.java_format.*;
import dev.argon.plugin.api_v1.util.ExIterable;
import org.jetbrains.annotations.NotNull;

import java.io.InputStream;

public interface ArgonModuleSerialized extends AutoCloseable {
    @NotNull Metadata metadata() throws Exception;
    @NotNull ModuleReferencesList references() throws Exception;

    @NotNull ExIterable<NamespaceDeclaration> namespaces() throws Exception;
    @NotNull ExIterable<GlobalDeclarationElement> namespaceElements(int id) throws Exception;

    @NotNull TraitDefinition getTraitDef(int id) throws Exception;
    @NotNull TraitReference getTraitRef(int id) throws Exception;

    @NotNull ClassDefinition getClassDef(int id) throws Exception;
    @NotNull ClassReference getClassRef(int id) throws Exception;

    @NotNull DataConstructorDefinition getDataConstructorDef(int id) throws Exception;
    @NotNull DataConstructorReference getDataConstructorRef(int id) throws Exception;

    @NotNull FunctionDefinition getFunctionDef(int id) throws Exception;
    @NotNull FunctionReference getFunctionRef(int id) throws Exception;
    @NotNull InputStream getFunctionPayload(int id, String platformId) throws Exception;

    @NotNull MethodDefinition getMethodDef(int id) throws Exception;
    @NotNull MethodReference getMethodRef(int id) throws Exception;
    @NotNull InputStream getMethodPayload(int id, String platformId) throws Exception;

    @NotNull ClassConstructorDefinition getClassConstructorDef(int id) throws Exception;
    @NotNull ClassConstructorReference getClassConstructorRef(int id) throws Exception;
    @NotNull InputStream getClassConstructorPayload(int id, String platformId) throws Exception;
}
